[Continuation] Merge EmbeddedLLM/vllm-rocm into vLLM main

tjtanaa opened the pull request:
**Add ROCm- Support**

- [x] Dynamic code path selection for CUDA or ROCm in PyTorch
- [x] Llama2 support
- [x] SqueezeLLM ROCm
- [x] add documentation amd-installation.rst. Describing how to setup vLLM ROCm version.
- [x] format.sh all the code
- [x] Prepare amd.Dockerfile

As there are too many changes has been made after https://github.com/vllm-project/vllm/pull/1749 ,
the previous PR https://github.com/vllm-project/vllm/pull/1749 is closed as and continued here.


PR Authors:
@kliuae
@iAmir97
@tjtanaa 
@tanpinsiang

Contributer:
@pcmortiz

This pull request also incorporates the work from [Port most vLLM kernels to ROCm #1313](https://github.com/vllm-project/vllm/pull/1313 ) by @pcmoritz, which was not merged. We appreciate @pcmoritz's contribution.

pcmoritz committed:
port dtype_float16.cuh and cache_kernels.cu

pcmoritz committed:
port dtype_bfloat16.cuh

pcmoritz committed:
port attention_utils.cuh

pcmoritz committed:
port more kernels

pcmoritz committed:
fix typo

pcmoritz committed:
add cuda_compat.h

pcmoritz committed:
Merge branch 'main' into port-to-rocm

pcmoritz committed:
sync branches

pcmoritz committed:
update

pcmoritz committed:
update

pcmoritz committed:
fixes

pcmoritz committed:
cleanup

pcmoritz committed:
update

pcmoritz committed:
update

pcmoritz committed:
update

pcmoritz committed:
fmt

pcmoritz committed:
cleanup

pcmoritz committed:
refactor

pcmoritz committed:
update

pcmoritz committed:
Merge branch 'main' into port-to-rocm

iAmir97 committed:
detecting rocm and adding flag for compiling

iAmir97 committed:
using asm volatile instead of hip api

iAmir97 committed:
using asm volatile for type casting of f16

kliuae committed:
Hipifying csrc file to accomodate rocm builds

tjtanaa committed:
Checked CUDA ROCm Compatibility (#15)

Features

* Auto-code path selection
* support llama2
* support squeezellm rocm
* add documentation amd-installation.rst. Describing how to setup vllm ROCm version
* format.sh all the code
* add base amd.Dockerfile

---------

Co-authored-by: tjtanaa <tunjian.tan@embeddedllm.com>
Co-authored-by: kuanfu <kuanfu.liu@embeddedllm.com>

kliuae committed:
merged with latest upstream

kliuae committed:
format code

kliuae committed:
downgrade torch requirement in toml to torch 2.0.1 to accommodate ROCm support

kliuae committed:
Merged changes from vllm main

kliuae committed:
Merged with changes in vllm main

hongxiayang commented:
Can you move this pull request to draft mode if not yet ready for review? Thanks.

kliuae committed:
Updated Dockerfile, rocm installation guide and setuppy

casper-hansen commented:
Do you have any plans to add ROCm support for the AWQ GEMM kernel? I know it is particularly challenging to convert the PTX code.

WoosukKwon commented:
@casper-hansen We discouraged them to add AWQ kernel support, since it requires significant engineering and we will integrate the new Triton kernel (hopefully next week).

kliuae committed:
Updated amd installation guide and dockerfile

kliuae committed:
Added num_gpus for ray init in ROCm

kliuae committed:
Synced torch version with vllm main in pyproject.toml

kliuae committed:
Format code

kliuae committed:
Merge branch 'main' into vllm-cuda-rocm-dev

kliuae committed:
Updated dockerfile.rocm and requirements-rocm.txt

kliuae committed:
Disable mistral for ROCm

kliuae committed:
Format code

kliuae committed:
Revert to cuda kernels

kliuae committed:
Merge remote-tracking branch 'pcmoritz/port-to-rocm'

Co-authored-by: Philipp Moritz <pcmoritz@gmail.com>
Co-authored-by: Amir Balwel <amoooori04@gmail.com>

kliuae committed:
Port latest kernels to ROCm

kliuae committed:
Update readme

WoosukKwon commented:
@tjtanaa Thanks again for the amazing work! Please ping me whenever this PR is ready for review!

kliuae committed:
Cleaned up kernel code

tjtanaa commented:
> @tjtanaa Thanks again for the amazing work! Please ping me whenever this PR is ready for review!

Thank you. @WoosukKwon It is ready to be reviewed.

hongxiayang reviewed (approved):
Thanks for the amazing work. 
I just verified the Dockerfile.rocm and ran the benchmarking on llama2-7b model on MI210. The thruput is:
```
Throughput: 0.89 requests/s, 424.90 tokens/s
```
Some minor comments will follow.

hongxiayang commented on Dockerfile.rocm:
```diff
@@ -0,0 +1,63 @@
+ENV PATH=$PATH:/opt/rocm/bin:/libtorch/bin:
+ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/rocm/lib/:/libtorch/lib:
+ENV CPLUS_INCLUDE_PATH=$CPLUS_INCLUDE_PATH:/libtorch/include:/libtorch/include/torch/csrc/api/include/:/opt/rocm/include/:
+ENV PYTORCH_ROCM_ARCH=gfx900;gfx906;gfx908;gfx90a;gfx1030;gfx1101
```
gfx900 can be removed from the list. 
Is gfx1101 tested?

kliuae replied:
Pytorch is already built in the pulled `rocm/pytorch` base image so probably don't need to overwrite this env variable. Thank you for pointing this out.

hongxiayang commented on docs/source/getting_started/amd-installation.rst:
```diff
@@ -0,0 +1,80 @@
+Build from source with docker
+-----------------
+
+You can also build and install vLLM from source:
```
Probably remove the word "also", as it looks like this is following another installation approach. Or remove this sentence.

tjtanaa replied:
Thank you. We have updated the documentations. Please take a look.

hongxiayang commented on docs/source/getting_started/amd-installation.rst:
```diff
@@ -0,0 +1,80 @@
+
+.. _build_from_source:
+
+Build from source with docker
```
Maybe "Option 1: Installation using Docker" ?

tjtanaa replied:
Thank you. We have updated the documentations. Please take a look.

hongxiayang commented on docs/source/getting_started/amd-installation.rst:
```diff
@@ -0,0 +1,80 @@
+       vllm-rocm \
+       bash
+
+If you plan to install vLLM-ROCm on a local machine or start from a fresh docker image (e.g. pytorch+rocm5.7), you can follow the steps below:
```
Option 2: Step by Step Installation Instruction

tjtanaa replied:
Thank you. We have updated the documentations. Please take a look.

hongxiayang commented on docs/source/getting_started/amd-installation.rst:
```diff
@@ -0,0 +1,80 @@
+       -v <path/to/model>:/app/model \
+       vllm-rocm \
+       bash
+
```
There is an additional step that I adapted from the v0.2.1.post1-rocm branch, which is to run the run_benchmark.sh script available in that branch (it need to run "mkdir dataset" in /app directory before running the script.
It might be worth mentioning here after start container step.

tjtanaa replied:
We are planning to incorporate the steps from v0.2.1.post1-rocm branch into the [documentations](https://docs.vllm.ai/en/latest/) in the coming PRs.
@WoosukKwon What do you think?

hongxiayang commented on docs/source/getting_started/amd-installation.rst:
```diff
@@ -0,0 +1,80 @@
+.. _installation:
```
Might worth changing the main README.md to summarize the work you have done to support vllm on ROCM, and direct readers to this file.

tjtanaa replied:
Thank you for the suggestion. We are working on this.

WoosukKwon commented on vllm/model_executor/layers/quantization/squeezellm.py:
```diff
@@ -114,9 +114,17 @@ def apply_weights(self,
+                                    dtype=torch.float)
+            ops.squeezellm_gemm(reshaped_x, qweight, out_float, lookup_table)
+            out = out_float.to(dtype=torch.float16)
+            # do something specific for HIP
```
nit: Move to L118?

WoosukKwon commented on vllm/engine/ray_utils.py:
```diff
@@ -73,7 +75,12 @@ def initialize_cluster(
+        if torch.version.hip:
+            ray.init(address=ray_address,
+                     ignore_reinit_error=True,
+                     num_gpus=parallel_config.world_size)
```
A dumb question: Why do we need this?

WoosukKwon commented on vllm/model_executor/layers/attention.py:
```diff
@@ -160,6 +160,8 @@ def forward(
                 p=0.0,
                 scale=self.scale,
+                op=xops.fmha.MemoryEfficientAttentionFlashAttentionOp[0] if
+                (torch.cuda.is_available() and torch.version.hip) else None,
```
BTW, why do we need `torch.cuda.is_available` here and in other places? In which case can this be False?

WoosukKwon commented on vllm/model_executor/model_loader.py:
```diff
@@ -38,6 +38,16 @@
     "YiForCausalLM": YiForCausalLM,
 }
 
+# Models to be disabled in ROCm
+_ROCM_DISABLED_MODELS = [
+    # ROCm's flash attention does not support sliding window attention
+    # in models such as Mistral
+    "MistralForCausalLM",
+]
```
Do you happen to know when the sliding window will be supported by AMD? If it's not gonna happen in the near future, what about supporting the model without the sliding window (and print a warning msg to the users)? While this will effectively limit the maximum context length of Mistral to 4096, I think many users can still use this model. This doesn't need to be fixed in this PR. I'm just wondering your thoughts.

tjtanaa replied:
The AMD will not support flash-attention v2.3 with sliding windows in near future, as stated in https://github.com/ROCmSoftwarePlatform/flash-attention/issues/22#issuecomment-1831947643 
Sure. we can support the model without the sliding windows (and print a warning msg to the users).

WoosukKwon commented on vllm/model_executor/model_loader.py:
```diff
@@ -38,6 +38,16 @@
+    # in models such as Mistral
+    "MistralForCausalLM",
+]
+if torch.version.hip:
```
Instead of directly using this, can we define a function like `is_hip()` in `vllm/utils.py`?

WoosukKwon commented on vllm/engine/arg_utils.py:
Following the comment on AWQ, can we keep this file unchanged, but raises an error message for unsupported load formats and quantization methods in `_verify_load_format` and `_verify_quantization` methods of `ModelConfig`?

WoosukKwon commented on Dockerfile.rocm:
```diff
@@ -0,0 +1,63 @@
+RUN python3 -m pip install --upgrade pip
+RUN python3 -m pip install --no-cache-dir ray[all]
+
+CMD ["/bin/bash"]
```
```suggestion
CMD ["/bin/bash"]

```

WoosukKwon commented on setup.py:
```diff
@@ -200,8 +271,14 @@ def read_readme() -> str:
+    else:
+        with open(get_path("requirements.txt")) as f:
+            requirements = f.read().strip().split("\n")
+    print("requirements: ", requirements)
+    # exit()
```
Do you need to keep this?
```suggestion
```

WoosukKwon reviewed (commented):
Hi @tjtanaa, thanks for the amazing work! The code is super clean and well organized! I've left a preliminary review. Will add more after actually running it on AMD GPUs. Thanks again for the great work!

kliuae committed:
Added wrapper for setting devFuncAttributeMaxDynamicSharedMemorySize

kliuae committed:
Added wrapper for setting devFuncAttributeMaxDynamicSharedMemorySize

kliuae committed:
Updated ROCm warp size

tjtanaa commented on csrc/attention/attention_kernels.cu:
```diff
@@ -538,9 +542,10 @@ __global__ void paged_attention_v2_reduce_kernel(
 
+#ifndef USE_ROCM
 #define LAUNCH_PAGED_ATTENTION_V1(HEAD_SIZE)                                                  \
   cudaFuncSetAttribute(                                                                       \
```
Yes. Thank you for the suggestion, we will aset a wrapper for this CUDA API.

tjtanaa commented on csrc/cuda_utils_kernels.cu:
`cudaGetDevice` and `cudaDeviceGetAttribute` are compatible with AMD but are not needed.
We are porting into ROCm for consistency with the upstream.

tjtanaa commented on requirements-rocm.txt:
Sure. Currently, we are trying to keep the files structure consistent with existing codebase.

kliuae committed:
Format code

kliuae committed:
Check hip from wrapper

kliuae committed:
Format code

kliuae committed:
Enable support for mistral models

kliuae committed:
Fixed hip device attribute

kliuae committed:
Format code

kliuae committed:
Restored awq file

kliuae committed:
Format code

kliuae committed:
Merge latest vllm main

kliuae commented on csrc/attention/attention_kernels.cu:
```diff
@@ -23,7 +23,11 @@
+#ifndef USE_ROCM
 #define WARP_SIZE 32
+#else
+#define WARP_SIZE 64
```
To our knowledge the warp size of the MI200s (gfx9) supported in this port are 64, but may be different in other archs like in gfx10 where 32 is the default but 64 can also be used. In light of this we have updated the code so the warp size is decided by hip runtime which uses the default warp size unless configured otherwise.

WoosukKwon replied:
Thanks! Awesome!

kliuae committed:
Updated rocm dockerfile

kliuae committed:
Update amd installation guide

tjtanaa committed:
Update vLLM Documentations (#18)

Update the vLLM installation procedures on AMD platform.
Update vLLM documentations.

hongxiayang commented:
> @hongxiayang This is my benchmark result on llama-7b and ShareGPT (`benchmark_throughput.py`), which is quite different from your results but seems more reasonable.
> 
> GPU	A100	MI210x
> TFLOPs (FP16)	312	181
> Memory capacity	80 GB	64 GB
> Memory bandwidth	1.9 TB/s	1.6 TB/s
> Throughput	8.30 reqs/s	5.25 reqs/s

Wow, Is this one gpu, or 8 GPUs?  your number is quite different from mine, and I am wondering whether we had the same parameters when running the test, like `INFO 12-05 22:40:38 llm_engine.py:73] Initializing an LLM engine with config: model='/app/model', tokenizer='/app/model', tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.float16, max_seq_len=2048, download_dir=None, load_format=auto, tensor_parallel_size=1, quantization=None, seed=0)`

WoosukKwon reviewed (approved):
@tjtanaa Thanks again for the great work! I found the code super clean and well organized. I also like the detailed documentation and the provided docker image. I could run vLLM on MI210 very smoothly and the performance was great! Thanks a lot for the contribution.

Left some minor comments on the code style. Please take a look!

WoosukKwon commented:
@hongxiayang It's LLaMA2-7B on a single MI210x. Basically, it should be the same setup as yours.

hongxiayang commented:
> @hongxiayang It's LLaMA2-7B on a single MI210x. Basically, it should be the same setup as yours.
There might be some difference between MI210 and MI210x. Would you mind sharing the exact command/steps you used? I will re-run it with your command.

kliuae committed:
Updated setup.py, vllm/utils.py and amd-installation doc

kliuae committed:
Updated setup.py

kliuae committed:
Format code

kliuae committed:
Merge branch 'main' into vllm-cuda-rocm-mod

kliuae commented on vllm/utils.py:
```diff
@@ -27,10 +27,14 @@ def reset(self) -> None:
     """Returns the maximum shared memory per thread block in bytes."""
     # https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html
-    cudaDevAttrMaxSharedMemoryPerBlockOptin = 97
+    cudaDevAttrMaxSharedMemoryPerBlockOptin = 97 if (not is_hip()) else 74
```
They were there for clarity, but perhaps not very pythonic so we have drop them. Thank you for pointing this out!

tjtanaa commented on setup.py:
```diff
@@ -8,27 +8,81 @@
 # Supported NVIDIA GPU architectures.
-SUPPORTED_ARCHS = {"7.0", "7.5", "8.0", "8.6", "8.9", "9.0"}
+NVIDIA_SUPPORTED_ARCHS = {"7.0", "7.5", "8.0", "8.6", "8.9", "9.0"}
+ROCM_SUPPORTED_ARCHS = {"gfx90a", "gfx908", "gfx906", "gfx1030", "gfx1100"}
```
The way we compiled this list of rocm supported archs is based on what AMD is supporting for ROCm and HIP, furthermore each arch has its own set of assembly instructions we have to make sure the currently used assembly  instructions is supported by those archs as well.

To the best of our knowledge, the following are the ARCH requirements needed by different libraries:
1. Pytorch `gfx900 gfx906 gfx908 gfx90a gfx1030 gfx1101`
2. vLLM Custom Ops: `gfx90a gfx908 gfx906 gfx1030 gfx1100`
3. Flash-Attention-ROCm: `gfx90a gfx940 gfx941 gfx942`

Should we use the intersection of all three ARCH requirements instead?

WoosukKwon replied:
@tjtanaa Thanks for the detailed explanation. Sorry, I have little background on this stuff. Maybe I should learn more about ROCm and AMD GPUs 😂

As far as I understand, the vLLM custom ops support every "recent" AMD GPUs, and currently the support is limited by the ROCm Flash Attention. Is this correct?

tjtanaa replied:
We believe in near future, the supported GPU ARCH is going to be restricted by Flash Attention ROCm.

tjtanaa commented:
> @tjtanaa Thanks again for the great work! I found the code super clean and well organized. I also like the detailed documentation and the provided docker image. I could run vLLM on MI210 very smoothly and the performance was great! Thanks a lot for the contribution.
> 
> Left some minor comments on the code style. Please take a look!

@WoosukKwon We have done updating the code style. and replied to your questions regarding to the supported ARCHs.

WoosukKwon commented:
@tjtanaa @kliuae LGTM! Many thanks again for the wonderful work! This will be HUGE!!
